src/PlcServer.cs

```diff
@@ -43,7 +45,209 @@ public PlcServer(OpcPlcConfiguration config, PlcSimulation plcSimulation, TimeSe
+        }
+        catch (Exception ex)
+        {
+            Meters.RecordTotalErrors(nameof(CreateMonitoredItems), ex.GetType().ToString());
+            _logger.LogError(ex, "Error creating monitored items");
+            throw;
+        }
+    }
+
+    public override ResponseHeader Publish(
+        RequestHeader requestHeader,
+        SubscriptionAcknowledgementCollection subscriptionAcknowledgements,
+        out uint subscriptionId,
+        out UInt32Collection availableSequenceNumbers,
+        out bool moreNotifications,
+        out NotificationMessage notificationMessage,
+        out StatusCodeCollection results,
+        out DiagnosticInfoCollection diagnosticInfos)
+    {
+        try
+        {
+            OperationContext context = ValidateRequest(requestHeader, RequestType.CreateSubscription);
+
+            var responseHeader = base.Publish(requestHeader, subscriptionAcknowledgements, out subscriptionId, out availableSequenceNumbers, out moreNotifications, out notificationMessage, out results, out diagnosticInfos);
+
+            int events = 0;
+            int dataChanges = 0;
+            int diagnostics = 0;
+            notificationMessage.NotificationData.ForEach(x =>
+            {
+                if (x.Body is DataChangeNotification changeNotification)
+                {
+                    dataChanges += changeNotification.MonitoredItems.Count;
+                    diagnostics += changeNotification.DiagnosticInfos.Count;
+                }
+                else if (x.Body is EventNotificationList eventNotification)
+                {
+                    events += eventNotification.Events.Count;
+                }
+                else
+                {
+                    Console.WriteLine("Unknown notification type");
+                }
+            });
+
+            Meters.AddPublishedCount(context.SessionId.ToString(), subscriptionId.ToString(), dataChanges, events);
+
+            _logger.LogDebug("{function} successfully with session: {sessionId} and subscriptionId: {subscriptionId}.",
+                nameof(Publish),
+                context.SessionId,
+                subscriptionId);
+
+            return responseHeader;
+        }
+        catch (Exception ex)
+        {
+            Meters.RecordTotalErrors(nameof(Publish), ex.GetType().ToString());
+            _logger.LogError(ex, "Error publishing.");
+            throw;
+        }
+    }
+
+    public override ResponseHeader Read(
+        RequestHeader requestHeader,
+        double maxAge,
+        TimestampsToReturn timestampsToReturn,
+        ReadValueIdCollection nodesToRead,
+        out DataValueCollection results,
+        out DiagnosticInfoCollection diagnosticInfos)
+    {
+        try
+        {
+            var responseHeader = base.Read(requestHeader, maxAge, timestampsToReturn, nodesToRead, out results, out diagnosticInfos);
+
+            _logger.LogDebug("{function} completed successfully.", nameof(Read));

```

Comment:
>. [](http://example.com/codeflow?start=63&length=1)

Nit: For logs, we don't use a period/dot at the end of a sentence, we only do that in exceptions.

Reply:
fixed